Fix intra-node IP message fragmentation due to mismatched MTU between devices

joestringer opened the pull request:
```release-note
Fix issue where MTU-sized packets received the outside world would be fragmented within the node and dropped by L4+ policy. Existing endpoints must be re-deployed to fix the issue. For more information, see https://cilium.link/err-low-mtu .
```

Previously, we would set the MTU of endpoints and Cilium devices to 1450 by default, even if the underlying fabric connectivity would support up to 1500B. This was done to attempt to force everything to use an MTU that would fit inside tunnel packets encapsulated and sent over the device with access to the outside world. 

However, one unforseen side effect of this is that when a 1500B IP packet is sent into the node (outward-facing device with MTU 1500) from the world, and it is routed through to the `cilium_host` device with MTU 1450, the node itself will fragment such messages into multiple IP packets before the traffic hits Cilium. Despite the node theoretically being able to handle such larger (1500B) packets, the packets are unnecessarily fragmented. As a result, when latter fragments hit the policy handling part of the datapath, we are unable to extract the port properly (as the packet does not contain it), and L4-L7 policies cannot be applied to the packets. As a result, they are dropped. For web services this looks like a hang.

This PR proposes to raise the MTU of all devices to 1500, while introducing a default route in each endpoint that lowers the MTU to 1450. The intention of this is to retain the lower MTU behaviour for packet transmission outside the endpoint, while allowing larger packets to be received.

~~The PR also adds a check in the IPv4 paths so that when policy is enabled, all fragments are dropped with an obvious reason (rather than performing a lookup with a nonsense port).~~
_Dropped for now as it's an orthogonal change._

Don't be too scared of the large LoC change, it's mostly an upgrade of the `github.com/vishvananda/netlink` library version which adds support for route MTU.

Todo:
* [x] Clean CI run
* [x] Test upgrade
* [x] Test multi-node
* [x] Document upgrade
  * [x] Set up link for `https://cilium.link/err-low-mtu`

Potential followups:
* Add some kind of MTU API for Cilium to allow other parts such as scripts or plugins to query (or set?) the MTU so that everything interacting with the Cilium daemon can consistently use the same MTU  (would help to support higher MTU cases)
* Look at raising the MTU of all devices and creating a high-MTU route for local destinations.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4323)
<!-- Reviewable:end -->

joestringer committed:
pkg: Add MTU package

This package will hold variables useful for configuring MTU.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
node: Configure route MTUs depending on destination

Configure MTU for routes such that routes to local destinations will
have MTU 65535, whereas routes to other nodes use the tunnel MTU.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Configure MTU for devices using pkg/mtu

Rather than configuring all devices to use MTU 1450 to restrict
transmission of larger packets (to ensure packets fit neatly inside
tunnel encapsulation), the previous commit set a default route inside
each endpoint to restrict the MTU to 1450.

This allows us to raise the default MTU of all of the devices to 1500B
to match common settings with the physical device on the node, allowing
us to receive 1500B IP frames from the outside world and not need to
fragment them within the node when transmitting on the `cilium_host` or
endpoint veth devices.

Use the pkg/mtu-configured MTU for all devices (or default to 1500B for
generating default CNI configuration).

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
vendor: Update netlink library for route MTU

The newer version of vishvananda's netlink library allows route MTU to
be configured, and we need this for upcoming commits where we wish to
manage MTU for traffic destined to tunnel devices via routes. Update the
dependency to gain access to this functionality.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on pkg/mtu/mtu.go:
```diff
@@ -0,0 +1,114 @@
+		attrs := link.Attrs()
+
+		skip := true
+		for _, prefix := range deviceMTUPrefixes {
```
Moving this loop into a function will make the skip logic simpler and easier to understand

tgraf commented on pkg/node/manager.go:
```diff
@@ -191,6 +192,12 @@ func replaceNodeRoute(ip *net.IPNet) {
+	// If the route includes the local address, then the route is for
+	// local containers and we can use a high MTU for transmit. Otherwise,
+	// it needs to be able to fit within the MTU of tunnel devices.
+	if !ip.Contains(local) {
```
Nice, this is beautiful

joestringer replied:
If only it worked as intended ;-)

joestringer replied:
I retained this, as routes to non-local destinations should actually use the lower MTU. It doesn't affect endpoint MTU, however.

tgraf reviewed (dismissed):
This looks awesome!

joestringer committed:
docs: Describe MTU changes from v1.0 to v1.1

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
cni: Configure default route MTU in endpoints

This configures the default route inside CNI endpoints with an MTU that
is derived from the CNI configuration's MTU, minus the tunnel overhead
necessary to encapsulate packets to other nodes over, eg, VXLAN.

This provides us freedom to raise the MTU of the host device to allow
larger packets to be received (eg, from the outside world) while
ensuring that the endpoint will never send packets that are too large to
fit in a tunnel.

Signed-off-by: Joe Stringer <joe@covalent.io>
